### PR TITLE
alternative expandvars + tests

### PR DESCRIPTION
**Pull request by laerus**
This changes the expandvars implementation to use regular expressions thus becoming faster.
The regex now matches cases like those mentioned in #385 but this doesn't actual fix this issue. Any ideas on how the requested behavior could be implemented would be appreciated, although this PR could just serve as a refactor of `expandvars` with added tests.

### CODE REVIEW

File: xonsh/tools.py [hunk 1]
```diff
@@ -42,6 +42,13 @@
                             ON_LINUX, ON_WINDOWS, PYTHON_VERSION_INFO)
 
 
+# regular expressions for matching enviroment variables
+# i.e $FOO, ${'FOO'}
+posix_envvar_regex = re.compile(r'\$({(?P<quote>[\'\"])|)(?P<envvar>\w+)((?P=quote)}|(?:\1\b))')
```
Comment (melund) [inline xonsh/tools.py]: You should make these lazy. See the concept here: https://github.com/laerus/xonsh/blob/2dc7f2b5c322ddb640eacd7d814c48371cc6c445/xonsh/tools.py#L213

Comment (melund) [inline xonsh/tools.py]: I also think we usually give them upper case names

Comment (laerus) [inline xonsh/tools.py]: i was just checking lazyasd thinking about doing this !

**Comment by melund** [comment]
I haven't got time to really look at it now. But I like the negative diff :)

**Comment by laerus** [comment]
tried to make the regex string with comments about each part inside a raw triple quoted string while escaping newlines, didn't work out.

**Comment by laerus** [comment]
there is still a bug from not using `_get_env_string' to detype the value of envvar to string, more tests incoming

**Comment by scopatz** [comment]
Is this ready for another look @laerus?

**Comment by laerus** [comment]
yes! i will add a news entry.

File: xonsh/tools.py [hunk 2]
```diff
@@ -1434,121 +1435,41 @@ def check_for_partial_string(x):
...
+    Unknown variables are left unchanged."""
     env = builtins.__xonsh_env__
     if isinstance(path, bytes):
         path = path.decode(encoding=env.get('XONSH_ENCODING'),
                            errors=env.get('XONSH_ENCODING_ERRORS'))
     elif isinstance(path, pathlib.Path):
         # get the path's string representation
         path = str(path)
-    if '$' not in path and (not ON_WINDOWS or '%' not in path):
-        return path
```
Comment (scopatz) [inline xonsh/tools.py]: The above two lines should come back, I think

**Comment by scopatz** [comment]
Thanks for putting this in @laerus!  I like the general idea,  but I think that this adds a regression.  On Windows, you are able to mix $VAR and %VAR% and this looks like it requires %VAR% on windows.  I think that the regex needs to be a little more involved to handle this.

**Comment by laerus** [comment]
the second `if` that checks for '$' in path runs on both linux and windows independent  of the first `if` that only runs on windows, i prolly should add a test for windows that checks if both envvar syntaxes, in the same path, get expanded.

**Comment by laerus** [comment]
hmm no idea how to start AppVeyor on this. @scopatz

**Comment by melund** [comment]
@learus If you login to appweyor with your github  account you should have access to restart it . @xonsh/core-developers are added as administrators on the appweyor account. If it doesn't work I can add you directly as I did with scopatz

**Comment by laerus** [comment]
hey @melund ! I can't make sense of AppVeyor interface atm. I logged in with my github account but there was no entry under *projects* and when i added my fork of xonsh i can't seem to make it test only this branch, even though i choose to build only this on settings

**Comment by laerus** [comment]
managed to try this branch on a windows machine, now works as expected and all tests pass, something like this `'$foo %spam%'` on windows will be expanded in `'bar eggs'` @scopatz

**Comment by laerus** [comment]
@melund when i push commits in this branch AppVeyor doesn't even care, only travis run tests here, am i doing something wrong?

**Comment by melund** [comment]
No. Maybe I configured it wrong way. I will look at it later to night.

File: xonsh/tools.py (hunk 2 above)
Comment (scopatz) [inline xonsh/tools.py]: Ahh nevermind your logic is good.

**Comment by scopatz** [comment]
> managed to try this branch on a windows machine, now works as expected and all tests pass, something like this '$foo %spam%' on windows will be expanded in 'bar eggs' @scopatz

Ahh you are right! Mea culpa

**Comment by scopatz** [comment]
Thanks a ton!

**Pull request merged**